ee/clickhouse/queries/trends/total_volume.py

```diff
@@ -38,48 +33,156 @@ def _total_volume_query(self, entity: Entity, filter: Filter, team_id: int) -> T
+        params["date_to"] = format_ch_timestamp(filter.date_to, filter)
+        params["interval"] = filter.interval
+        return final_query, params, self._parse_total_volume_result(filter)
 
     def _parse_total_volume_result(self, filter: Filter) -> Callable:
         def _parse(result: List) -> List:
-            parsed_results = []
-            for _, stats in enumerate(result):
-                parsed_result = parse_response(stats, filter)
-                parsed_results.append(parsed_result)
+            # NOTE: there's some implicit coupling here:
+            #
+            #        1. the ordering of the fields we get back, (date, value)
+            #        2. that the last row is always the `WITH TOTALS` result
+            #
+            #  TODO: remove coupling
+
+            #  The last row relates to the `WITH TOTALS` from the clickhouse query
+            values_without_totals, (_, aggregated_value) = result[:-1], result[-1]
+
+            # Separate dates and values, as desired for the REST API response
+            dates, values = zip(*sorted(values_without_totals))
+
+            parsed_result = parse_response([dates, values], filter)
 
-            return parsed_results
+            # NOTE: to maintain backwards compatability here from previous behaviour, we exclude the
+            #        `aggregated_value` for some filter.display.
+            # TODO: verify if updating to always including `aggregated_value` would break anything. It
+            #       would be simpler to not have to have this special casing in here.
+            if filter.display in TRENDS_DISPLAY_BY_VALUE:
+                parsed_result["aggregated_value"] = aggregated_value
+
+            return [parsed_result]
 
         return _parse
+
+
+ZERO_FILL_TEMPLATE = """
+SELECT 
+    interval.start AS interval_start,
+
+    -- For where we don't have an aggregate value, we zero fill
+    COALESCE(aggregate.value, 0) AS value
+
+FROM (
+    -- Creates zero values for all date axis ticks for the given date_from, date_to range
+    -- NOTE: I had a look at using the `WITH FILL` modifier for `ORDER BY` but it didn't work
+    --       well for week and month intervals. Looks like you'd have to do some other wrangling
+    --       to get it to work as expected:
+    --
+    --           https://stackoverflow.com/questions/66092272/clickhouse-order-by-with-fill-month-interval
+    --
+
+    SELECT {{interval}}(
+        toDateTime(%(date_to)s) - {{interval_func}}(number)
+    ) AS start
+
+    FROM numbers(
+        dateDiff(
+            %(interval)s, 
+            {{interval}}(toDateTime(%(date_from)s)), 
+            {{interval}}(toDateTime(%(date_to)s))
+        ) + 1
+    )
+) interval
+
+-- left join so we end up with values for all intervals, even if we don't have an aggregate
+LEFT JOIN (
+        
+    {aggregate_query}
+
+) aggregate ON aggregate.interval_start = interval_start
+"""
+
+
+def zero_fill(aggregate_query: str) -> str:
```

Comment:
This introduces a new querying pattern:

1. Query is stored in the same file
2. We use functions instead of `.format` calls to add subqueries

I'd push back on this unless there's a deeper reason here - having too many querying patterns makes code comprehension harder.

Reply:
I’m not a fan of just using format here. It’s interface is not obvious (the required args are implicitly specified in the string which makes it tricky to comprehend).

Using a function here I think allows for a bit more clarity, communicates that this is something that can be reused (more so than an arbitrary string var like null_sql for which it is not obvious that I should be coupling my code to), specify the cases that it should be used.

Also happy to move to format